Framework/HistogramData/test/HistogramIteratorTest.h

```diff
@@ -0,0 +1,412 @@
+  }
+
+  void test_iterator_decrement() {
+    Histogram hist(Points{1, 2, 3}, Frequencies{2, 3, 4});
+    auto iter = hist.end();
+    --iter;
+    TS_ASSERT_DIFFERS(iter, hist.begin());
+    TS_ASSERT_EQUALS(iter->frequency(), 4);
+    --iter;
+    TS_ASSERT_DIFFERS(iter, hist.begin());
+    TS_ASSERT_EQUALS(iter->frequency(), 3);
+    --iter;
+    TS_ASSERT_EQUALS(iter, hist.begin());
+    TS_ASSERT_EQUALS(iter->frequency(), 2);
+  }
+
+  void test_iterator_advance() {
+    Histogram hist(Points{1, 2, 3}, Frequencies{2, 3, 4});
+    auto iter = hist.begin();
+
+    std::advance(iter, 2);
+    TS_ASSERT_EQUALS(iter->frequency(), 4);
+    // check past end of valid range
+    std::advance(iter, 2);
+    TS_ASSERT_EQUALS(iter, hist.end());
+    std::advance(iter, -3);
+    TS_ASSERT_EQUALS(iter->frequency(), 2);
+    // check before start of valid range
+    std::advance(iter, -2);
+    TS_ASSERT_EQUALS(iter, hist.begin());
+  }
+
+  void test_iterator_distance() {
+    Histogram hist(Points{1, 2, 3}, Frequencies{2, 3, 4});
+    auto begin = hist.begin();
+    auto end = hist.end();
+
+    TS_ASSERT_DIFFERS(begin, end);
+    TS_ASSERT_EQUALS(std::distance(begin, end), 3);
+    ++begin;
+    TS_ASSERT_DIFFERS(begin, end);
+    TS_ASSERT_EQUALS(std::distance(begin, end), 2);
+    --begin;
+    std::advance(begin, std::distance(begin, end) / 2);
+    TS_ASSERT_DIFFERS(begin, end);
+    TS_ASSERT_EQUALS(std::distance(begin, end), 2);
+  }
+
+  void test_iterate_over_histogram_counts() {
+    // Arrange
+    Counts expectedCounts{2, 3, 4};
+    Histogram hist(Points{1, 2, 3}, expectedCounts);
+
+    // Act
+    auto result =
+        std::equal(hist.begin(), hist.end(), expectedCounts.begin(),
+                   [](const HistogramItem &item, const double &counts) {
+                     return item.counts() == counts;
+                   });
+
+    // Assert
+    TSM_ASSERT("Counts did not match", result);
+  }
+
+  void test_iterate_over_histogram_counts_when_histogram_has_frequencies() {
+    // Arrange
+    Histogram hist(Points{1, 2, 3}, Frequencies{2, 3, 4});
+    Counts expectedCounts = hist.counts();
+
+    // Act
+    auto result =
+        std::equal(hist.begin(), hist.end(), expectedCounts.begin(),
+                   [](const HistogramItem &item, const double &counts) {
+                     return item.counts() == counts;
+                   });
+
+    // Assert
+    TSM_ASSERT("Counts did not match", result);
+  }
+
+  void test_iterate_over_histogram_frequencies() {
+    // Arrange
+    Frequencies expectedFrequencies{2, 3, 4};
+    Histogram hist(Points{1, 2, 3}, expectedFrequencies);
+
+    // Act
+    auto result =
+        std::equal(hist.begin(), hist.end(), expectedFrequencies.begin(),
+                   [](const HistogramItem &item, const double &frequency) {
+                     return item.frequency() == frequency;
+                   });
+
+    // Assert
+    TSM_ASSERT("Frequencies did not match", result);
```

Comment:
Also just a minor (matter of taste?) suggestion: You have a test that tests only a single thing, so the test name alone gives a good indication as to what may have gone wrong. Personally I would simply write

```TS_ASSERT(std::equal(...));```

For me that makes clear what is going on, and if you have a failure you get the name of the failed test from ctest, telling you what went wrong. No need to write fancy assert messages, I think.